go/pserver/service.go

```diff
@@ -81,30 +78,78 @@ type Service struct {
 	optMap             map[string]*optimizer
 }
 
+// parameterCheckpoint saves parameter checkpoint
+type parameterCheckpoint struct {
+	ParameterWithConfig
+	State []byte
+}
+
+// NewCheckpointFromFile loads parameters and state from checkpoint file
+func NewCheckpointFromFile(cpPath string, idx int, e *EtcdClient) (*Checkpoint, error) {
+	v, err := e.GetKey(PsPath+string(idx), 3)
+	if err != nil {
+		return nil, err
+	}
+
+	var cpMeta checkpointMeta
+	if err = json.Unmarshal(v, &cpMeta); err != nil {
+		return nil, err
+	}
+
+	fn := filepath.Join(cpPath, cpMeta.UUID)
+	if _, err = os.Stat(fn); os.IsNotExist(err) {
+		return nil, err
+	}
+
+	f, err := os.Open(fn)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	h := md5.New()
+	if _, err = io.Copy(h, f); err != nil {
+		return nil, err
+	}
+	md5 := hex.EncodeToString(h.Sum(nil))
+	if md5 != cpMeta.MD5 {
+		return nil, errors.New(CheckpointUnmatched)
+	}
+
+	dec := gob.NewDecoder(f)
```

Comment:
Since `io.Copy` already read through `f` once, `gob.NewDecoder(f)` requires to read from the beginning of the file, but `f` have not been `seek` to the beginning of the file. Does this actually work?